### ISSUE TITLE
Format truncation

### ISSUE THREAD

**Issue by leahneukirchen** [Yubico/libfido2#318]
**What version of libfido2 are you using?**

1.7.0

**What operating system are you running?**

Void Linux, x86_64 glibc

**How does the problem manifest itself?**

GCC fails to build official releases due to `-Werror`.

**Is the problem reproducible?**

Yes

**What are the steps that lead to the problem?**

Compile with GCC 10.2.1

```
../src/log.c:41:37: warning: '%s' directive output may be truncated writing up to 127 bytes into a region of size between 2 and 254 [-Wformat-truncation=]
```

**Comment by martelletto** [comment]
Hi,

Thank you for the report.

I don't seem to be able to reproduce the problem. I am new to Void Linux, so I am probably doing something wrong. Here's what I am doing:

```
$ VOID="$(docker run -it -d voidlinux/voidlinux:latest)"                        
$ docker exec -it $VOID xbps-install -Su                                        
$ docker exec -it $VOID xbps-install cmake gcc-10.2.1pre1_3 git libcbor-devel libudev-devel make openssl-devel pkg-config zlib zlib-devel
$ docker exec -it $VOID git clone https://github.com/yubico/libfido2            
$ docker exec -it $VOID /bin/sh -c 'cd libfido2 && cmake -B build .'            
$ docker exec -it $VOID make -C /libfido2/build 
```

I am using GCC 10.2.1, x86_64, and glibc, as far as I can tell:

```
$ docker exec -it $VOID cc --version
cc (GCC) 10.2.1 20201203
$ docker exec -it $VOID arch                  
x86_64
$ docker exec -it $VOID ls -l /usr/lib64/libc.so.6  
lrwxrwxrwx 1 root root 12 Jan 21 12:00 /usr/lib64/libc.so.6 -> libc-2.32.so
```

-p.

**Comment by martelletto** [comment]
Note to self: the problem is triggered by a combination of CFLAGS="-O3 -D_FORTIFY_SOURCE=2" (which are enabled by CMAKE_BUILD_TYPE=Release), gcc-8 or gcc-10, and glibc headers. The involvement of -D_FORTIFY_SOURCE is clear from the full error log:

```
/libfido2/src/log.c: In function 'fido_log_xxd':
/libfido2/src/log.c:41:37: error: '%s' directive output may be truncated writing up to 127 bytes into a region of size between 2 and 254 [-Werror=format-truncation=]
   41 |   snprintf(line, sizeof(line), "%s: %s\n", body, suffix);
      |                                     ^~
......
   80 |  do_log(row, fmt, args);
      |         ~~~                          
In file included from /usr/include/stdio.h:866,
                 from /libfido2/src/log.c:10:
/usr/include/bits/stdio2.h:70:10: note: '__builtin___snprintf_chk' output between 4 and 383 bytes into a destination of size 256
   70 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   71 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/libfido2/src/log.c: In function 'fido_log_error':
/libfido2/src/log.c:41:37: error: '%s' directive output may be truncated writing up to 255 bytes into a region of size between 2 and 254 [-Werror=format-truncation=]
   41 |   snprintf(line, sizeof(line), "%s: %s\n", body, suffix);
      |                                     ^~
......
  110 |  do_log(errstr, fmt, args);
      |         ~~~~~~                       
In file included from /usr/include/stdio.h:866,
                 from /libfido2/src/log.c:10:
/usr/include/bits/stdio2.h:70:10: note: '__builtin___snprintf_chk' output between 4 and 511 bytes into a destination of size 256
   70 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   71 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

```

**Comment by martelletto** [comment]
Fixed in 8edb9a204b2f4aeb487e282908c3187f1d02d606. Users running into this problem while building 1.7.0 are encouraged to set `-Wno-error=format-truncation` in CFLAGS prior to calling cmake.
